### ISSUE TITLE
Distributed workers

### ISSUE THREAD

**Issue by Lex-au** [gpustack/gpustack#664]
Hi Team,

Can you provide some advice on how to distribute load across multiple machines? I have two machines in resources. I can distribute inference to one machine, or the other; but I cannot get both machines to share inference.

I have tried using binpack, and assigning labels to both machines. These labels work if I want to isolate a single machine to run inference, but trying to use both machine labels results in pending "no suitable workers".

**Comment by gitlawr** [comment]
>  I cannot get both machines to share inference.

I don't get it. Could you elaborate on what models you’re using, what GPUs are available and how do you intend to run?

**Comment by gitlawr** [comment]
Does [this tutorial](https://docs.gpustack.ai/latest/tutorials/performing-distributed-inference-across-workers/) help?

**Comment by Lex-au** [comment]
> > I cannot get both machines to share inference.
> 
> I don't get it. Could you elaborate on what models you’re using, what GPUs are available and how do you intend to run?

I have two machines connected over a 10Gbps LAN. One is a headless workstation equipped with two NVIDIA A5000 GPUs (48GB combined VRAM) and 128GB of RAM, while the other is a workstation with an RTX 4090 (24GB VRAM) and 64GB of RAM. This gives me a total of 72GB VRAM across all three GPUs.

I’m running Mistral Large 123b via LlamaBox into an API endpoint for OpenWebUI. Currently, I can run a replica from my 4090 workstation, which deploys the model to the dual A5000s, offloading some to CPU/RAM when needed. However, I can't seem to get it to avoid offloading to the CPU. Additionally, I want to fully utilise the network to distribute inference back across the 4090 GPU.

Instead of leveraging the full 72GB VRAM across all three GPUs, the system is only using the A5000 machine's 48GB VRAM over the network. While I can use labels to force inference on either machine, I can't get both machines to share the distribution of inference workloads.

I’ve tried looking into modifying the docker-compose.yml to designate nodes for distributed inference, but according to the repository documentation, this step shouldn’t be necessary as distributed inference should work out of the box.

Could this be a configuration issue with LlamaBox, or is there something I’m missing to enable proper VRAM distribution across GPUs and machines?

**Comment by gitlawr** [comment]
When possible, partial GPU offloading on a single node is preferred over distributed across multiple nodes due to networking overhead. The schedule priority is listed [here](https://docs.gpustack.ai/latest/scheduler/#resource-fit-policy). If you don't want CPU offloading, you can disable it in model configuration. Please check if that works for you. For example:
<img width="1091" alt="image" src="https://github.com/user-attachments/assets/34c35de4-57e2-41b3-b19e-e8882ace2454">

**Comment by Lex-au** [comment]
> When possible, partial GPU offloading on a single node is preferred over distributed across multiple nodes due to networking overhead. The schedule priority is listed [here](https://docs.gpustack.ai/latest/scheduler/#resource-fit-policy). If you don't want CPU offloading, you can disable it in model configuration. Please check if that works for you. For example: <img alt="image" width="1091" src="https://private-user-images.githubusercontent.com/5697937/391010296-34c35de4-57e2-41b3-b19e-e8882ace2454.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4Njc5NTEsIm5iZiI6MTczMjg2NzY1MSwicGF0aCI6Ii81Njk3OTM3LzM5MTAxMDI5Ni0zNGMzNWRlNC01N2UyLTQxYjMtYjE5ZS1lODg4MmFjZTI0NTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjlUMDgwNzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDdjNDBkYTU2ODQzNmMwMDU3MTY2MDhmN2E2OTNhZmNmMmZmNGVjZTFhNjkxOGU0NzU4YzlhYzM4ZWEyN2IwOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fZDaNjITaWEkjCt1azc47Dang8P8AnOZL9yGLLUgT2A">

I have use Ray cluster over 10Gbps without issue for similar size models, but I chose to try GPUstack as it claimed out of the box distribution. So I don't believe in this case it's network overhead. Though perhaps I have misconstrued what you meant.

Is there any foresable reason you can think of why this does not seem to work for me? I have disabled CPU offloading, without using labels to mark workers, and it results in a "pending - no suitable worker" result.

![image](https://github.com/user-attachments/assets/8b7669c6-e2b4-4f4e-82dc-962869d66380)
![image](https://github.com/user-attachments/assets/e641e4fc-f649-4795-b2dc-eec42ba20151)
![image](https://github.com/user-attachments/assets/4f80aa33-9994-4dee-9fa7-9c4f306f4908)
![image](https://github.com/user-attachments/assets/33eec0c2-9317-4dac-9cff-c0898f94ef79)

**Comment by Lex-au** [comment]
To provide further context, a hugging face model I can "run" distributed across both nodes; however, I believe its a false positive because there's no VRAM usage on either machine and the endpoint sits idling without inference to input. I am wondering if perhaps I am fundamentally missing something here.

![image](https://github.com/user-attachments/assets/bb841a59-929e-4bc0-a8c2-adbd1d1db2af)

**Comment by gitlawr** [comment]
Can you request the inference apis to see how the resource usage goes? You can also compare with the nvidia-smi output to see check if the the metrics(with a few latency) match.

**Comment by Lex-au** [comment]
> ~Can you request the inference apis to see how the resource usage goes? You can also compare with the nvidia-smi output to check if the the metrics(with a few latency) match.~
> 
> I missed the previous comment. Let me do some calculation.

That's more than okay. I appreciate the assistance. I am sure this is likely something simple I have fundamentally misunderstood.

**Comment by gitlawr** [comment]
This is essentially how it works,
it uses [gguf-parser](https://github.com/gpustack/gguf-parser-go) to estimate the resource requirement for distributed deployment similar to the following:
```shell
$ gguf-parser -hfr bartowski/Mistral-Large-Instruct-2411-GGUF -hff Mistral-Large-Instruct-2411-Q2_K_L.gguf --ctx-size 8192 --tensor-split 24,24,24 --rpc h1:p1

+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| ESTIMATE                                                                                                                                                                                                                                                                                                                               |
+-------+--------------+--------------------+-----------------+-----------+----------------+-------------+---------------+----------------+----------------+----------------------------------------------+----------------------------------------------+--------------------------------------+----------------------------------------+
|  ARCH | CONTEXT SIZE | BATCH SIZE (L / P) | FLASH ATTENTION | MMAP LOAD | EMBEDDING ONLY |  RERANKING  | DISTRIBUTABLE | OFFLOAD LAYERS | FULL OFFLOADED |                      RAM                     |                 RPC 0 (V)RAM                 |                VRAM 0                |                 VRAM 1                 |
|       |              |                    |                 |           |                |             |               |                |                +--------------------+------------+------------+----------------+--------------+--------------+----------------+---------+-----------+----------------+-----------+-----------+
|       |              |                    |                 |           |                |             |               |                |                | LAYERS (I + T + O) |     UMA    |   NONUMA   | LAYERS (T + O) |      UMA     |    NONUMA    | LAYERS (T + O) |   UMA   |   NONUMA  | LAYERS (T + O) |    UMA    |   NONUMA  |
+-------+--------------+--------------------+-----------------+-----------+----------------+-------------+---------------+----------------+----------------+--------------------+------------+------------+----------------+--------------+--------------+----------------+---------+-----------+----------------+-----------+-----------+
| llama |     8192     |     2048 / 512     |     Disabled    |  Enabled  |       No       | Unsupported |   Supported   |   89 (88 + 1)  |       Yes      |      1 + 0 + 0     | 240.88 MiB | 390.88 MiB |     30 + 0     |   15.14 GiB  |   17.02 GiB  |     30 + 0     | 1.01 GB | 17.02 GiB |     28 + 1     | 14.53 GiB | 16.41 GiB |
+-------+--------------+--------------------+-----------------+-----------+----------------+-------------+---------------+----------------+----------------+--------------------+------------+------------+----------------+--------------+--------------+----------------+---------+-----------+----------------+-----------+-----------+

```

If the result cannot fit current setup, the scheduling tells no suitable workers for the deployment. I am not sure which quantization is used here. You can provide the exact model info for us or you can try with the above command to see if it's expected or a false possitive decision.

We are also working on making the message clear but the estimation is kinda complicated at the moment.

**Comment by pengjiang80** [comment]
For  the "false positive" distributed problem, did you see VRAM utilization when deploying a fully GPU offloaded model or a CPU/GPU mixed model in a single worker? 

If not, could you help to check if correct llama-box file was downloaded to your workers? You can use the following command to get the path to the file. The please help to check the file size.  We have different llama-box binaries which will be downloaded based on hardware detection automatically. 
```shell
ps -ef | grep llama-box
```

**Comment by Lex-au** [comment]
Thanks for the assistance team. To provide further context, I am running this natively without WSL (so purely llamabox, not VLLM). Though I could try deploying your docker version, but at that point, I would simply return to use ray cluster.

I think it may ultimately be something I've missed, though today I will try manually setting nodes in the yaml files and see if I can workaround it.

I'll update here once done.

**Comment by Lex-au** [comment]
Issue resolved.

**Comment by pengjiang80** [comment]
Hi Lex, thanks for your update. How did you finally get it resolved? We want to confirm if there is any bug we should fix or any improvements we can make in GPUStack.

**Comment by Lex-au** [comment]
> Hi Lex, thanks for your update. How did you finally get it resolved? We want to confirm if there is any bug we should fix or any improvements we can make in GPUStack.

Hi Team,

Thank you for your response. Here's an update with further details:

Initially, I resolved the issue by editing the Docker Compose files to explicitly designate the nodes. However, I later discovered that it is possible to achieve a distributed workflow without modifying the Docker Compose setup or designating nodes manually. That said, turning off CPU offloading and switching the scheduler from spread to binpack remains necessary for GGUF models quantized for parallelization (designated by K) to function correctly. Though, if no one else has reported this, it may be solely related to my experience.

Additionally, I encountered a few issues that might be worth investigating:

**Desynchronization in Distributed Inference:**
When running distributed inference over a LAN without internet access, I observed occasional desynchronization and VRAM unloading after a period of time. This behavior does not occur when internet access is available. Is there any heartbeat functionality in the codebase that could explain this behavior?

**Replica Restart Hanging:**
In cases where desynchronization occurs, attempting to restart the replica leads to a hanging initialization stage, which remains pending and eventually results in an inference error code 0. Interestingly, with internet access, the system transitions past the initialization phase to perform a download check before running successfully.

These observations might provide useful insights for debugging or potential improvements. Let me know if I can assist further by investigating specific parts of the codebase.
